src/nnet3/nnet-simple-component.cc

```diff
@@ -668,12 +718,99 @@ void ClipGradientComponent::Backprop(const std::string &debug_info,
       in_deriv->ApplyCeiling(clipping_threshold_);
       in_deriv->ApplyFloor(-1 * clipping_threshold_);
     }
+
+    to_update->num_backproped_ += 1;
+    RepairGradients(debug_info, in_value, in_deriv, to_update);
   }
 }
 
+void ClipGradientComponent::RepairGradients(
+    const std::string &debug_info,
+    const CuMatrixBase<BaseFloat> &in_value,
+    CuMatrixBase<BaseFloat> *in_deriv, ClipGradientComponent *to_update) const {
+  KALDI_ASSERT(to_update != NULL);
+  BaseFloat default_clipped_proportion_threshold = 0.01;
+  BaseFloat default_clipping_to_value = 0.0;
+
+  // we use this 'repair_probability' (hardcoded for now) to limit
+  // this code to running on about half of the minibatches.
+  BaseFloat repair_probability = 0.5;
```

Comment:
>
> @danpovey <https://github.com/danpovey> I checked in a new commit with
> all changes you suggested except this one. I don't quite understand the use
> of repair_probability. According to your above comment, it seems relavant
> to the direction of derivatives. But in my understanding, the nonlinearity
> component's self-repair and self-repair here both try to add a term to
> in-deriv to force the input value towards 0, so an input with value < 0 and
> an input with value > 0 move in opposite directions in both cases.
>

I guess you're right.  Maybe it's really no different.

However, after thinking about all this, I'm getting increasingly unhappy
with a certain aspect of it.
Doing the self-repair will increase the magnitudes of the gradients; and
we're measuring the magnitudes of the gradients to decide whether to do
self-repair.
My concern is that if, for a certain job, we start doing self-repair, it
can become self-perpetuating for the lifetime of that job, because the
self-repair would increase the magnitudes of the gradients and lead to more
clipping, hence more self-repair.
I'm not sure how to fix this.  But maybe it's not really a problem.  See
how consistent the amount of self-repair is between jobs.  If some do much
more than others, maybe it's a problem.